### DISCUSSION TITLE
How to format large number?

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#6583] [General]
I like using either `,` or `_` to display large numbers and it is quite straightforward to do in PostgreSQL using the `to_char` data type formatting functions ([doc link](https://www.postgresql.org/docs/current/functions-formatting.html)). Here is a small PostgreSQL example query:

```sql
with t(i,n) as (select i,10^i from generate_series(1,9) g(i)) 
select *,to_char(n::int,'999G999G999G999') expected from t;
```
which yields

```
┌───┬────────────┬──────────────────┐
│ i │     n      │     expected     │
├───┼────────────┼──────────────────┤
│ 1 │         10 │               10 │
│ 2 │        100 │              100 │
│ 3 │       1000 │            1,000 │
│ 4 │      10000 │           10,000 │
│ 5 │     100000 │          100,000 │
│ 6 │    1000000 │        1,000,000 │
│ 7 │   10000000 │       10,000,000 │
│ 8 │  100000000 │      100,000,000 │
│ 9 │ 1000000000 │    1,000,000,000 │
└───┴────────────┴──────────────────┘
(9 rows)
```

How can I generate the `expected` column like format in duckdb? I tried `format(n, '{:,f}')` thinking it was like Python's format string [^python-example] but kept on getting weird errors (`Error: Invalid Error: missing '}' in format string`).

Having `_` separators for large numbers is hugely helpful! It would be awesome if duckdb also allows inputting large numbers with this kind of a separator. Many newer languages (Go, Python, rust) already allow this kind of input and it would be super helpful if duckdb can also do this. I'm assuming this would require changing how number parsing is done and I hope it's not too difficult to do this. By the way, I think the upcoming PostgreSQL version is also going to allow this kind of number input ([ref](https://www.depesz.com/2023/02/06/waiting-for-postgresql-16-allow-underscores-in-integer-and-numeric-constants/))!

Thanks in advance!

[^python-example]: Try the following in the python repl to see what I mean: `print(f"{1000:,f} x {1000:_d} = {1000*1000:,d}")` .

**Comment by l1t1** [comment]
I tried the example of the doc. and add :.2f
```
D select format('Benchmark "{}" took {} seconds', 'CSV', 42);
┌─────────────────────────────────────────────────────┐
│ format('Benchmark "{}" took {} seconds', 'CSV', 42) │
│                       varchar                       │
├─────────────────────────────────────────────────────┤
│ Benchmark "CSV" took 42 seconds                     │
└─────────────────────────────────────────────────────┘
D select format('Benchmark "{}" took {:.2f} seconds', 'CSV', 42);
Error: Invalid Error: precision not allowed for this argument type
```

**Reply by orzom411** [reply]
Maybe try:
`D select format('Benchmark "{}" took {:.2f} seconds', 'CSV', cast( 42 as numeric ));`
┌────────────────────────────────────────────────────────────────────────────────┐
│ format('Benchmark "{}" took {:.2f} seconds', 'CSV', CAST(42 AS DECIMAL(18,3))) │
│                                    varchar                                     │
├────────────────────────────────────────────────────────────────────────────────┤
│ Benchmark "CSV" took 42.00 seconds                                             │
└────────────────────────────────────────────────────────────────────────────────┘

42 was an int hence the argument type disagreement?  At least that was how I read it.

**Comment by l1t1** [comment]
the fmt of format() should be well documented.

**Reply by vlulla** [reply]
You might find the tests provided in the source helpful... https://github.com/duckdb/duckdb/blob/master/test/sql/function/string/test_format.test . HTH.

**Comment by l1t1** [comment]
the feature is too simple
